Refactor to use Workbook Pattern

KJoslyn opened the pull request:
Converted project selection, chapter selection, and verse selection to use workbook pattern.

TakeManagementViewModel.kt is in the process of a major refactor and can be ignored, since take management functionality was already broken.

Requires [common #112](https://github.com/WycliffeAssociates/otter-common/pull/112)

KJoslyn committed:
Create function onChangeAndDoNow in TornadoFXUtils.kt

KJoslyn committed:
Convert activeProject to Workbook

project display and selection process is converted to Workbook pattern

KJoslyn committed:
Convert activeCollection to Chapter

KJoslyn committed:
Convert activeContent to Chunk

TakeManagementViewModel is no longer functional

KJoslyn committed:
More descriptive names for fragments

KJoslyn committed:
ResourceListViewModel automatically loads resources

KJoslyn committed:
Load resources from source chapter

The current workbook pattern assumes that we navigate through the target book chapters. But resources need to be loaded from the source book

KJoslyn committed:
Fix broken build

KJoslyn commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resources/viewmodel/ResourceListViewModel.kt:
```diff
@@ -6,16 +6,30 @@ import org.wycliffeassociates.otter.jvm.app.ui.resourcetakes.viewmodel.RecordRes
+        }
+    }
+
+    private fun getSourceChapter(targetChapter: Chapter): Chapter {
```
As of now, activeChapter and activeChunk in WorkbookViewModel correspond to the `target` Book in the workbook. But the resources are found in the `source` book. Not sure if we want to handle this differently.

aunger replied:
That seems correct to me.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/CardDataMapper.kt:
```diff
@@ -1,36 +1,26 @@
-            }
-            return cardList
-        }
+        fun mapChunkToCardData(chunk: Chunk) = CardData(
```
Suggestion (disregard if you think this is better): These two methods seem like they would be very natural secondary constructors on CardData.

KJoslyn replied:
I like that, so I made the change. If @sarabiaj prefers the old way, though, I can very easily revert it- it was a 1 minute change.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/CardDataMapper.kt:
```diff
@@ -1,36 +1,26 @@
+            CardDataType.CONTENT.value,
+            chunk.start.toString(),
+            chunk.sort,
+            chunkSource = chunk
```
Suggestion: Use named parameters for all of these args.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/viewmodel/CardGridViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+
+    // Whether the UI should show the plugin as active
+    private var showPluginActive: Boolean by property(false)
+    val showPluginActiveProperty = getProperty(CardGridViewModel::showPluginActive)
```
These are unused, and are maybe out of place?

KJoslyn replied:
Not sure why they were there. I renamed this file but git didn't recognize it as just a rename apparently. They do appear to be out of place, though- there is a property with the same name in TakeManagementViewModel (in later PR's it is in RecordableViewModel since TMVM got deleted). I just removed this property from CardGridViewModel since it is unused here.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/viewmodel/CardGridViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+                    if (chapterModeEnabledProperty.value == true) {
+                        allContent.filtered { it.item == ContentLabel.CHAPTER.value }
+                    } else {
+//                        allContent.filtered { it.item != ContentLabelEnum.CHAPTER.value }
```
Remove dead code.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/viewmodel/CardGridViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+    val chapterModeEnabledProperty = SimpleBooleanProperty(false)
+
+    init {
+        Observable.merge(chapterModeEnabledProperty.toObservable(), allContent.changes()).subscribe { _ ->
```
line length

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/viewmodel/CardGridViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+        workbookViewModel.activeChapterProperty.onChange {
+            it?.let { ch -> loadChapterContents(ch) }
+                ?: workbookViewModel.activeWorkbookProperty.value
+                    ?.let { wb -> loadChapters(wb) }
```
how about:
```
?: workbookViewModel.activeWorkbookProperty.value?.let { 
    wb -> loadChapters(wb) 
}
```

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/viewmodel/CardGridViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+            it?.let { ch -> loadChapterContents(ch) }
+                ?: workbookViewModel.activeWorkbookProperty.value
+                    ?.let { wb -> loadChapters(wb) }
+        }
```
This is a little hard for me to read, and I think it's because you're using what appears to be a transformation chain but is actually a decision about which side-effect (Unit-return) function to call.  Maybe better to set this us with `if/else` or `when`.

Suggestion: Also, to my eyes, it's better to name the variable earlier than later, and use Kotlin's smart casts, instead of (e.g.) using `it` for a little while, then `ch` for the same value.

jsarabia replied:
I ended up just calling out the formatting but had the same feeling, and thought if/else would be better too

KJoslyn replied:
Switched to `when`, it definitely is more readable.

KJoslyn committed:
Remove nullable from WorkbookViewModel properties

Properties are already nullable, so we don't need to specify the properties as nullable explicitly

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/viewmodel/CardGridViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+            .toList()
+            .doOnSuccess {
+                allContent.setAll(it)
+            }.subscribe()
```
Replace doOnSuccess + subscribe with
```
            .subscribe { list: List<CardData> ->
                allContent.setAll(list)
            }
```
Something is funny about certain Single.subscribe overloads, so I needed to specify the type explicitly, and maybe this is why you used `doOnSuccess`.

jsarabia replied:
@KJoslyn what are your thoughts here?

KJoslyn replied:
It was a while ago, but that may have been why I did it that way (there wouldn't be any other reason.) I just changed this and the other instance of this in the same file.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/viewmodel/CardGridViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+                // setAll is causing the UI to hang, probably because node structure is complex. If "loading" is
+                // set to false after this operation, the spinner will remain but stop spinning while the UI hangs.
+                allContent.setAll(it)
+            }.subscribe()
```
Same comment as above re `doOnSuccess`

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/cardgrid/viewmodel/CardGridViewModel.kt:
```diff
@@ -0,0 +1,103 @@
+            .doOnSuccess {
+                // TODO
+                // setAll is causing the UI to hang, probably because node structure is complex. If "loading" is
+                // set to false after this operation, the spinner will remain but stop spinning while the UI hangs.
```
Is it any better to drop the `toList` and add individual elements?

KJoslyn replied:
No, actually if you do that, you'll see one or a few chapters show up in the grid, and then the UI hangs for a very long time before all the rest of the chapters show up. Not sure what gives.

KJoslyn committed:
PR Comments

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/mainscreen/viewmodel/MainScreenViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+        if(takesPageDocked.value == false) {
+            find<MainScreenView>().activeFragment.dock<TakeManagementFragment>()
+        }
+        takesPageDocked.set(true)
```
Move into `if`?

KJoslyn replied:
I wonder if this was just a holdover from some previous version of the code, and if that previous version observed `takesPageDocked` somewhere. (Not sure why else it would be a property.) Anyway, `takesPageDocked` doesn't do anything useful, since the only time the value is read is before deciding whether to dock the `TakeManagementFragment`. But trying to dock a fragment that is already docked just results in a no-op anyway. So I actually just removed `takesPageDocked.`

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/takemanagement/viewmodel/TakeManagementViewModel.kt:
```diff
@@ -204,18 +202,18 @@ class TakeManagementViewModel : ViewModel() {
-                        EditTake.Result.NO_EDITOR -> snackBarObservable.onNext(messages["noEditor"])
-                    }
-                }
+//        contextProperty.set(TakeContext.EDIT_TAKES)
```
why are you commenting this out?

KJoslyn replied:
@sarabiaj @aunger I'm going to go ahead and just resolve all of the comments about TakeManagementViewModel since the entire file got refactored and removed in a later PR anyway.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/takemanagement/viewmodel/TakeManagementViewModel.kt:
```diff
@@ -261,7 +259,9 @@ class TakeManagementViewModel : ViewModel() {
         title = if (activeContentProperty.value?.type == ContentType.META) {
-            activeCollectionProperty.value?.titleKey ?: ""
+            // TODO
+            "TODO"
```
don't display TODO in the app; it'll likely stay that way

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/mainscreen/viewmodel/MainScreenViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private fun setActiveChunkText(chunk: Chunk) {
+        selectedChunkTitle.set(ContentLabel.of(chunk.contentType).value.toUpperCase())
```
Internationalization / look up in Messages?

KJoslyn replied:
I just talked to Thomas about those nav box cards, and we're not entirely sure how the nav box cards are going to work now that selecting a resource actually selects both a chunk and a resource. UNLESS the resource is about the chapter as a whole, which is a different case. It seems to me like this issue has to be deferred until we know what's really going to happen.

jsarabia replied:
What? 
In the mean time could you just put it in a messages lookup?

KJoslyn replied:
That's fine, for now I just changed it to`   selectedChunkTitle.set(messages["verse"].toUpperCase())`

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/mainscreen/viewmodel/MainScreenViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private fun setActiveChapterText(chapter: Chapter) {
+        selectedChapterTitle.set(ContentLabel.CHAPTER.value.toUpperCase())
```
i18n

KJoslyn replied:
This one is easier, I just changed it to `selectedChapterTitle.set(messages["chapter"].toUpperCase())`

KJoslyn replied:
Maybe me writing this comment will help me remember i18n... (If I were put in detention I'd be writing this like 50 times haha...)

I won't forget i18n!!

aunger replied:
haha :)

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/resources/viewmodel/ResourceListViewModel.kt:
```diff
@@ -6,16 +6,30 @@ import org.wycliffeassociates.otter.jvm.app.ui.resourcetakes.viewmodel.RecordRes
-        val chapter = workbookViewModel.chapter
+    init {
+        workbookViewModel.activeChapterProperty.onChangeAndDoNow {
+            it?.let { targetChapter ->
```
Suggestion: Name `it` as `targetChapter` earlier and use smart-cast.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/takemanagement/viewmodel/TakeManagementViewModel.kt:
```diff
@@ -82,7 +79,8 @@ class TakeManagementViewModel : ViewModel() {
             populateTakes(it)
-            getContentList(activeCollection)
+            // TODO
+//            getContentList(activeCollection)
```
When will the TODO be resolved?  If not immediately, then at least make the TODO more descriptive.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/takemanagement/viewmodel/TakeManagementViewModel.kt:
```diff
@@ -204,18 +202,18 @@ class TakeManagementViewModel : ViewModel() {
+//                        }
+//                        EditTake.Result.NO_EDITOR -> snackBarObservable.onNext(messages["noEditor"])
+//                    }
+//                }
```
`// ???`

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/takemanagement/viewmodel/TakeManagementViewModel.kt:
```diff
@@ -261,7 +259,9 @@ class TakeManagementViewModel : ViewModel() {
-            activeCollectionProperty.value?.titleKey ?: ""
+            // TODO
+            "TODO"
+//            activeCollectionProperty.value?.titleKey ?: ""
```
More description needed here

jsarabia committed:
Update .travis.yml

Specify distro to be able to use oraclejdk8

KJoslyn committed:
Turn card data mapper into secondary constructors on CardData

KJoslyn committed:
PR Comments

KJoslyn committed:
Remove takesPageDocked property

KJoslyn committed:
PR Comments

KJoslyn committed:
i18n for Chunk navbox card

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/jvm/app/ui/mainscreen/viewmodel/MainScreenViewModel.kt:
```diff
@@ -54,7 +54,7 @@ class MainScreenViewModel : ViewModel() {
 
     private fun setActiveChunkText(chunk: Chunk) {
-        selectedChunkTitle.set(ContentLabel.of(chunk.contentType).value.toUpperCase())
+        selectedChunkTitle.set(messages["verse"].toUpperCase())
```
you couldn't do:
```selectedChunkTitle.set(messages[ContentLabel.of(chunk.contentType).value])```
?

KJoslyn replied:
Probably. Something makes me uncomfortable about the underlying assumptions about ContentType and ContentLabel that we know should result in `messages["verse"]` in this case. But we did just do a refactor on these classes, and as long as they are not changing it's probably fine. Sorry if I'm making this a bigger deal than it is!

KJoslyn committed:
dynamic i18n for Chunk navbox card
